src/libcore/slice.rs

```diff
@@ -677,6 +677,8 @@ macro_rules! iterator {
             fn next(&mut self) -> Option<$elem> {
                 // could be implemented with slices, but this avoids bounds checks
                 unsafe {
+                    ::intrinsics::assume(!self.ptr.is_null());
+                    ::intrinsics::assume(!self.end.is_null());
```

Comment:
Is this duplicating the work of https://github.com/rust-lang/rust/pull/21886 as well? (maybe leave to #21886?)